src/NativeScript/Calling/FFICall.cpp

```diff
@@ -72,24 +76,101 @@ void FFICall::visitChildren(JSCell* cell, SlotVisitor& visitor) {
-        void* argumentBuffer = reinterpret_cast<void**>(buffer + this->_argsArrayOffset)[i + this->_initialArgumentIndex];
-        JSCell* parameterType = this->_parameterTypesCells[i].get();
-        this->_parameterTypes[i].write(execState, argument, argumentBuffer, parameterType);
+    EncodedJSValue JSC_HOST_CALL FFICall::call(ExecState* execState) {
+        FFICall* callee = jsCast<FFICall*>(execState->callee());
+        Invocation invocation(*callee);
+        ReleasePoolHolder releasePoolHolder;
 
+        callee->preCall(execState, invocation);
+        callee->_invocationHooks.pre(callee, execState, invocation);
         if (execState->hadException()) {
-            return;
+            return JSValue::encode(execState->exception());
+        }
+
+        {
+            JSLock::DropAllLocks locksDropper(execState);
+            ffi_call(callee->_cif, FFI_FN(invocation.function), invocation._buffer + callee->_returnOffset, reinterpret_cast<void**>(invocation._buffer + callee->_argsArrayOffset));
+        }
+
+        JSValue result = callee->_returnType.read(execState, invocation._buffer + callee->_returnOffset, callee->_returnTypeCell.get());
+
+        if (InvocationHook post = callee->_invocationHooks.post) {
+            post(callee, execState, invocation);
         }
+
+        return JSValue::encode(result);
+    }
+
+    JSObject* FFICall::async(ExecState* execState, JSValue thisValue, const ArgList& arguments) {
+        Invocation* invocation = new Invocation(*this);
+        ReleasePoolHolder releasePoolHolder;
+
+        Register* fakeCallFrame = new Register[JSStack::CallFrameHeaderSize + execState->argumentCount() + 1];
+        ExecState* fakeExecState = ExecState::create(fakeCallFrame);
+        fakeExecState->init(nullptr, nullptr, CallFrame::noCaller(), arguments.size() + 1, this);
+        fakeExecState->setThisValue(thisValue);
+        fakeExecState->setCallerFrame(execState->callerFrame());
+        for (size_t i = 0; i < arguments.size(); i++) {
+            fakeExecState->setArgument(i, arguments.at(i));
+        }
+        ASSERT(fakeExecState->argumentCount() == arguments.size());
+
+        {
+            TopCallFrameSetter frameSetter(execState->vm(), fakeExecState);
+            this->preCall(fakeExecState, *invocation);
+            this->_invocationHooks.pre(this, fakeExecState, *invocation);
+            if (fakeExecState->hadException()) {
+                return fakeExecState->exception();
+            }
+        }
+
+        JSPromiseDeferred* deferred = JSPromiseDeferred::create(execState, execState->lexicalGlobalObject());
+        auto* releasePool = new ReleasePoolBase::Item(std::move(releasePoolHolder.relinquish()));
+
+        dispatch_async(dispatch_get_global_queue(0, 0), ^{
+            ffi_call(this->_cif, FFI_FN(invocation->function), invocation->_buffer + this->_returnOffset, reinterpret_cast<void**>(invocation->_buffer + this->_argsArrayOffset));
```

Comment:
I have a question. Could the `this` value be moved inside here. Also you could use the `invocation.resultBuffer()` method here.